### PR TITLE
#157963716 API endpoints 

### PR DESCRIPTION
**Pull request by RuthNjeri**
#### What does this PR do?
It creates the API tests.
#### Description of Task to be completed?
1. Write an endpoint for the  POST request `/maintenanceapp/api/v1/requests/`  to create a request
2. Write an endpoint for a PUT request `/maintenanceapp/api/v1/requests/<requestId>` to modify a request
3. Write an endpoint for a Get request `/maintenanceapp/api/v1/requests/` allowing users logged in to read their requests.
4. Write an endpoint for a specific GET request `/maintenanceapp/api/v1/requests/<requestId>` allowing a logged in user to fetch a request belonging to them.
5. Write an endpoint to DELETE a request allowing a user to delete a request belonging to them `/maintenanceapp/api/v1/requests/<requestId>`
#### How should this be manually tested?
First clone the repository
Navigate inside the Maintenance-Tracker-Application folder. `cd app` folder. open the `__init__.py`. Test the endpoints using [Postman](https://www.getpostman.com/)

#### What are the relevant pivotal tracker stories?
[#157963716](https://www.pivotaltracker.com/n/projects/2173304)

#### Screenshots
get_request_with_id
get_requests
newnewpost
put_request
deleterequest

### CODE REVIEW

File: .pytest_cache/v/cache/lastfailed [hunk 1]
```diff
@@ -0,0 +1,11 @@
+{
```
Comment (zmwas) [inline .pytest_cache/v/cache/lastfailed:1]: Add to gitignore

File: .pytest_cache/v/cache/nodeids [hunk 2]
```diff
@@ -0,0 +1 @@
+[]
```
Comment (zmwas) [inline .pytest_cache/v/cache/nodeids]: This too

File: app/__init__.py [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+
+#Initialize application
+# def create_app(config_name):
+#     app = FlaskAPI(__name__,instance_relative_config=True)
+#     app.config.from_object(app_config[config_name])
+#     app.config.from_pyfile('config.py')
+
+#     return app
+
+requests = [
```
Comment (zmwas) [inline app/__init__.py]: Don't hardcode the requests. They should be stored in memory not on your file

File: app/__init__.py [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+    return jsonify({'requests':requests})
+
+@app.route('/maintenanceTrackerApp/api/v1/requests/<int:request_id>', methods=['GET'])
+def get_request(request_id):
+    	request = [request for request in requests if request['id']==request_id]
+    	if len(request) == 0:
+    		abort(404)
+    	return jsonify({'request':request[0]})
+
+# @app.errorhandler(404)
```
Comment (zmwas) [inline app/__init__.py]: Uncomment this code

File: app/__init__.py [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+#!flask/bin/python
+
+
+from flask import Flask,request,jsonify,abort
+from flask import make_response
+# from instance.config import app_config
+
+app = Flask(__name__)
+
+#Initialize application
```
Comment (zmwas) [inline app/__init__.py]: Why is this commented out?

File: tests/.pytest_cache/v/cache/lastfailed [hunk 6]
```diff
@@ -0,0 +1,3 @@
+{
```
Comment (zmwas) [inline tests/.pytest_cache/v/cache/lastfailed:1]: Add this in your gitignore

File: .pytest_cache/v/cache/lastfailed (hunk 1 above)
Comment (RuthNjeri) [reply .pytest_cache/v/cache/lastfailed:1]: Alright.

File: app/__init__.py (hunk 3 above)
Comment (RuthNjeri) [reply app/__init__.py]: Working on it

File: app/__init__.py (hunk 5 above)
Comment (RuthNjeri) [reply app/__init__.py]: the from import errors

File: app/__init__.py [hunk 7]
```diff
@@ -1,12 +1,115 @@
...
+
+@app.route('/maintenanceTrackerApp/api/v1/requests', methods=['POST'])
+def create_request():   
+		if not request.json or not 'title' in request.json:
+			abort(400)
+		if len(requests)==0:
+			print(1)	
+			app_request = {
+
+						'id': request.json['id'],
```
Comment (zmwas) [inline app/__init__.py]: The json should not have the id when creating a request as the user does not know how many requests are available. Use something like `len(requests) +1`

File: app/__init__.py [hunk 8]
```diff
@@ -1,12 +1,115 @@
...
-from flask import Flask
 
+from flask import Flask,request,jsonify,abort
+from flask import make_response
+# from instance.config import app_config
 
-#Initialize application
-app = Flask(__name__,instance_relative_config=True)
+app = Flask(__name__)
+"""
```
Comment (zmwas) [inline app/__init__.py]: If the commented chunk of code is not being used delete it before pushing it.

File: app/__init__.py [hunk 9]
```diff
@@ -1,12 +1,115 @@
...
+						'description': request.json['description'],
+						'type':request.json['type']
+			
+						}
+			requests.append(app_request)
+		else:
+			print(2)	
+			app_request = {
+
+					'id': requests[-1]['id'] + 1,
```
Comment (zmwas) [inline app/__init__.py]: Why not use len(requests)+1 ? Is there a reason for this approach?

File: .pytest_cache/v/cache/nodeids (hunk 2 above)
Comment (RuthNjeri) [reply .pytest_cache/v/cache/nodeids]: Added

File: tests/.pytest_cache/v/cache/lastfailed (hunk 6 above)
Comment (RuthNjeri) [reply tests/.pytest_cache/v/cache/lastfailed:1]: Added

**Comment by zmwas** [comment]
For deleting a better response would be something like `"message":"Successfully deleted"`

File: app/__init__.py (hunk 8 above)
Comment (RuthNjeri) [reply app/__init__.py]: Noted.

File: app/__init__.py (hunk 9 above)
Comment (RuthNjeri) [reply app/__init__.py]: It is an if condition that if there are requests in the list, it picks the last id of the request..adds it to one and appends it to the group of requests.

File: app/views.py [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+        return make_response(jsonify({'error':'Not found'}),404)  
+
+@app.route('/maintenanceapp/api/v1/requests', methods=['POST'])
+def create_request():   
+        if not request.json or not 'title' in request.json:
+            abort(400)
+        if len(requests)==0:    
+            app_request = {
+
+                        'id': request.json['id']+1,
```
Comment (zmwas) [inline app/views.py]: The id is supposed to be auto-generated not passed in the request. Use len() and 
then add 1.

File: app/views.py [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+
+@app.errorhandler(404)
+def request_not_found(error):
+        return make_response(jsonify({'error':'Not found'}),404)  
+
+@app.route('/maintenanceapp/api/v1/requests', methods=['POST'])
+def create_request():   
+        if not request.json or not 'title' in request.json:
+            abort(400)
+        if len(requests)==0:
```
Comment (zmwas) [inline app/views.py]: The if statement is not needed as long as you use len(requests) and add 1 as the id.

File: app/views.py [hunk 12]
```diff
@@ -0,0 +1,142 @@
...
+        return jsonify({'message': 'Successfully deleted' })    
+
+
+@app.route('/maintenanceapp/api/v1/users/', methods=['POST'])
+def create_user():
+    if len(users)==0:
+        
+            app_request = {
+
+                        'id': request.json['id']+1,
```
Comment (zmwas) [inline app/views.py]: Same thing with this id.

File: app/views.py [hunk 13]
```diff
@@ -0,0 +1,142 @@
...
+        request = [request for request in requests if request['id']==request_id]        
+        if len(request) == 0:
+            abort(404)
+        request.remove(request[0])
+        return jsonify({'message': 'Successfully deleted' })    
+
+
+@app.route('/maintenanceapp/api/v1/users/', methods=['POST'])
+def create_user():
+    if len(users)==0:
```
Comment (zmwas) [inline app/views.py]: Same case with this if.

**Comment by zmwas** [comment]
Write tests for edge cases for a user signing up and logging in.

File: app/views.py (hunk 10 above)
Comment (RuthNjeri) [reply app/views.py]: Done

File: app/views.py (hunk 11 above)
Comment (RuthNjeri) [reply app/views.py]: ooh yeah. Changing that now.

File: app/views.py (hunk 12 above)
Comment (RuthNjeri) [reply app/views.py]: removed

File: app/views.py (hunk 13 above)
Comment (RuthNjeri) [reply app/views.py]: It has been changed

File: app/views.py (hunk 10 above)
Comment (zmwas) [reply app/views.py]: You havent' changed it

Comment (RuthNjeri) [reply app/views.py]: @zmwas I have pushed the changes now [commit](https://github.com/RuthNjeri/Maintenance-Tracker-Application/pull/11/commits/98d4f084574b8a3d0b4806e91a14f06c3ce7e324)
